### PR TITLE
[core][lua] Mob Base Damage Plumbing

### LINKED ISSUE #8174: 🐛 Mobs that use H2H only attack once instead of twice.
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my issue will be ignored.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have searched existing [issues](https://github.com/LandSandBoat/server/issues) to see if the issue has already been opened, and I have checked the commit log to see if the issue has been resolved since my server was last updated.

## Branch affected by issue 

`base`

## Steps to reproduce

Find a MNK mob like a Mandragora and aggro it. See that it only attacks once.

I think it was unintentionally introduced in https://github.com/LandSandBoat/server/pull/8080/ when a TYPE_PC check was added to CAttackRound::IsH2H().

@WinterSolstice8 

## Expected behavior

Mobs that use Hand to Hand should attack twice per round.

### PR DESCRIPTION
**Pull request by UmeboshiXI**
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

1. Resolves https://github.com/LandSandBoat/server/issues/8174
 * Moved the `objtype == TYPE_PC` check from `CAttackRound::IsH2H()` to `CAttackRound::CreateKickAttacks()` in the block that checks for MERIT_KICK_ATTACK_RATE to prevent crash when called by non PC.
 
2. Allow mobs to utilize kick attacks(Previously was players only). I observed this happening on normal MNK mobs on retail.

3. First draft of implementing closer to retail mob base damage using [Jimmayus's data](https://docs.google.com/spreadsheets/d/1YBoveP-weMdidrirY-vPDzHyxbEI2ryECINlfCnFkLI/edit?pli=1&gid=1743955268#gid=1743955268).

4. Implements mob H2H auto attack damage penalty using above data in link.

To summarize the data:

* Vanilla - COP Zones: Normal mob damage is Level + 2 offset (Currently what LSB uses)

* TOAU Zones: Level + 10 offset

* WOTG+: Level + 11 offset

* Some mob families have custom offsets.

* Mobs in beginner zones outside the 3 cities have an additional -1 offset compared to normal.

* Some mobs have a multiplier to the base damage value but before the offset.

* Some mobs such as Korrigans/Tauri have their offset included in damage multipliers. (Will be implemented by setting their offset mod to 0 and then adding a WEAPON_BONUS mobmod with the value of what the offset was.)

`BASE_DAMAGE_MULTIPLIER`: Used for things like Morbols which according to the spreadsheet have a 1.5x base damage multiplier before the offset is added.

`DAMAGE_OFFSET`: Allows us to set a custom offset. Potentially useful if edge cases come up (Think HNMs and their base damage).

`RANGED_DAMAGE_OFFSET`: Same as above but for Ranged slot

## Steps to test these changes

Go get hit by stuff in different expansion zones. Take note of MNK mobs especially.

MNK mobs on average should hit less hard per hit compared to current LSB.
Mobs in TOAU and WOTG zones should hit a bit harder.

Note: The mob mods are not currently used, a follow up PR will be done to go through and implement known values based off Jimmayus's data.

### CODE REVIEW

**Comment by WinterSolstice8** [comment]
you saw mobs using kick attacks? are you sure it wasn't just DA?

**Comment by UmeboshiXI** [comment]
> you saw mobs using kick attacks? are you sure it wasn't just DA?

I tested on a Level 67 Mourioche in B. Tree. They should be MNK/MNK.
Back was also turned so no counters.

image

**Comment by WinterSolstice8** [comment]
> I tested on a Level 67 Mourioche in B. Tree. They should be MNK/MNK.

Interesting! do you have cap for that? If not i'll go get one, REALLY curious as to what the packet says for the animation on the third attack

**Comment by UmeboshiXI** [comment]
Capture software is on Windower right(I'm on ashita atm)? Might need a bit to get everything set up unless you want to grab it real quick. I got 2~ procs in a couple minutes when I doubled checked just now.

**Comment by WinterSolstice8** [comment]
> Capture software is on Windower right(I'm on ashita atm)?

Sruon has _massively_ upgraded Captain and the old windower suite is obsolete. If you need any new features you can make an issue on his repo.

https://github.com/sruon/captain

Make sure to download the zip release

**Comment by UmeboshiXI** [comment]
Thanks, give me a little bit to get OBS and capture suit installed. I'll get out and retest.

File: src/map/attack.cpp [hunk 1]
```diff
@@ -657,10 +657,28 @@ void CAttack::ProcessDamage()
 
         if (m_attackType == PHYSICAL_ATTACK_TYPE::KICK)
```
Comment (WinterSolstice8) [inline src/map/attack.cpp]: I would move this case after the TYPE_MOB case, I don't see any reason why kicks would be lower base damage than an "unarmed" mob. There is a WoTG NM that exclusively uses Footwork but we'll get to that later

**Comment by UmeboshiXI** [comment]
Got a capture:

https://www.dropbox.com/scl/fi/4ovhdle2az9rq67tocbji/Kick-Attack-Testing.zip?rlkey=8z03p85vn0khkx2sqyx4b7ita&st=uuf1nx4q&dl=0

https://youtu.be/LA1_G_Iq2Wc?t=201

3 hit attack is around 15:28:00

**Comment by WinterSolstice8** [comment]
image

Yup, that's a real kick attack alright.  Uses animation 2 which is "right kick attack"

that's crazy. KA has been wrong for mobs since forever... good job on that find

File: src/map/attack.cpp (hunk 1 above)
Comment (UmeboshiXI) [reply src/map/attack.cpp]: So mob kick attack will use the same mob H2H calculation? If so should I make a kick attack check to add KICK_DMG mod back in?

Comment (WinterSolstice8) [reply src/map/attack.cpp]: Yeah. that sounds good. I don't know if we'll ever actually need the mod but better to have it

Comment (UmeboshiXI) [reply src/map/attack.cpp]: Force pushed an update. If this looks good, I'll go ahead and copy the changes over to the lua version of this attack function.

Comment (WinterSolstice8) [reply src/map/attack.cpp]: Yeah looks good, go ahead

**Comment by UmeboshiXI** [comment]
> image
> 
> Yup, that's a real kick attack alright. Uses animation 2 which is "right kick attack"
> 
> that's crazy. KA has been wrong for mobs since forever... good job on that find

Noob question. What is this interface in the picture?

**Comment by UmeboshiXI** [comment]
Converted from draft, should be good to go now.

**Review by WinterSolstice8 — APPROVED** [review]

File: scripts/enum/mob_mod.lua [hunk 2]
```diff
@@ -94,4 +94,10 @@ xi.mobMod =
     CLAIM_TYPE             = 83, -- Changes the claim behavior of the mob. See xi.claimType enum.
     NO_SPELL_COST          = 84, -- Mob does not use MP when casting spells
     ASTRAL_PET_OFFSET      = 85, -- If non-zero, defines the offset from main mob's ID for astral flow (if zero, will assume offset of 2)
+    BASE_DAMAGE_OVERRIDE   = 86, -- Overrides the mob's base damage.
+    BASE_DAMAGE_MULTIPLIER = 87, -- Multiplies the mob's base damage. This is applied BEFORE MOBMOD_BASE_DAMAGE_OFFSET. BASE 10000 (15000 = x1.5)
+    DAMAGE_MULTIPLIER_MODE = 88, -- Default: 0. Setting to 1 includes the damage offset in the MOBMOD_BASE_DAMAGE_MULTIPLIER. Used for special mob cases such as Korrigan and Tauri.
+    DAMAGE_OFFSET          = 89, -- Overrides the mob's base damage offset.
+    RANGED_DAMAGE_OFFSET   = 90, -- Overrides the mob's ranged base damage offset.
+    FAMILY_DAMAGE_OFFSET   = 91, -- Adjusts the existing offset after region/mod corrections (Some mob families have slight differences in offset).
```
Comment (Frankie-hz) [inline scripts/enum/mob_mod.lua]: What is the family mod going to be used for? If you take rabbits for example you cant apply the same mod to the family because its in every type of zone.

What is the difference between DAMAGE_OFFSET and WEAPON_BONUS? You wrote `DAMAGE_OFFSET: Allows us to set a custom offset. Potentially useful if edge cases come up (Think HNMs and their base damage).` but that is already what WEAPON_BONUS does. Should we just use one of them instead of having two mods that do the same thing?

**Comment by WinterSolstice8** [comment]
> Noob question. What is this interface in the picture?

https://github.com/ZeromusXYZ/VieweD

I was using an old version of that (too lazy to update shortcuts)

its good for spot checking stuff

File: scripts/enum/mob_mod.lua (hunk 2 above)
Comment (UmeboshiXI) [reply scripts/enum/mob_mod.lua]: Got clarification from Jimmayus. So far it seems like Rabbits/Worms in Vanilla/ROTZ have the -1 offset. I was thinking I could make a new family for the earlier rabbits/worms to use.

This could go for other mobs as well since apparently there are some mob families that have buffs up to COP but lose them in TOAU+. 

The alternative would be to add these mods into pool mods.

> What is the difference between DAMAGE_OFFSET and WEAPON_BONUS? You wrote DAMAGE_OFFSET: Allows us to set a custom offset. Potentially useful if edge cases come up (Think HNMs and their base damage). but that is already what WEAPON_BONUS does. Should we just use one of them instead of having two mods that do the same thing?

I guess you can functionally do the same thing with Weapon Bonus (Like in the recent Yovra PR). I personally did it this way to have the mod functions be explicit instead of having to math around WEAPON_BONUS. Open to feedback either way.

Also on that note, I noticed there was no underflow/overflow safety for WEAPON_BONUS/this function before(And by extension this PR). I'll work on that as well.

**Comment by WinterSolstice8** [comment]
I found a DA + Kick on a 50+ yagudo
image

and 2x DA
image

still waiting for 2x DA + kick...

**Comment by WinterSolstice8** [comment]
image

2x DA + kick
left fist, right fist, left fist, left fist, right kick

**Comment by UmeboshiXI** [comment]
Slightly off topic. Zanshin is another thing I'd like to test in future with packet capture. Some reports of it procing on mobskills with Lamia SAMs Tidal Slash skill.

**Comment by WinterSolstice8** [comment]
> Slightly off topic. Zanshin is another thing I'd like to test in future with packet capture. Some reports of it procing on mobskills with Lamia SAMs Tidal Slash skill.

Player WS 100% don't Zanshin (I used the PW 100% Zanshin gkt to test it in pvp) but that doesn't mean mobs can't
If it's a one hit WS you could check if it has better than floored hit rate when your evasion is capped vs the mob to check

**Comment by UmeboshiXI** [comment]
@WinterSolstice8 Made an adjustment to mobutils.cpp. Mind looking over it one last time when you have time?

File: scripts/enum/mob_mod.lua (hunk 2 above)
Comment (Xaver-DaRed) [reply scripts/enum/mob_mod.lua]: If I may give my 2 cents, 6 different mobmods to override damage is simply overkill. It will make stuff impossible to track, specially if we start mixing stuff from families aswell.

The number should be 3, at most:
- Regular attack offset.
- Ranged attack offset.
- attack multiplier. No mode. If its 0-> its a 1. If it isnt 0, its the value.

And then we should just adapt to the limitations, just so we can keep this sane, even if its inconvenient at times.

EDIT: Thinking better, just offset should do nicely and be easier than a full override.

Comment (UmeboshiXI) [reply scripts/enum/mob_mod.lua]: I'll strip it down. How would you suggest I handle this:

```
        // MOBMOD_DAMAGE_MULTIPLIER_MODE determines which damage formula we use for multiplying.
        if (PMob->getMobMod(MOBMOD_DAMAGE_MULTIPLIER_MODE) == 1)
        {
            // Special cases(Korrigan/Tauri mobs) add the offset BEFORE multiplying.
            damage = (damage + offset) * multiplier;
        }
        else
        {
            // Normal mobs typically add the offset AFTER multiplying.
            damage = damage * multiplier + offset;
        }
```

Since certain mobs have a different multiplier calculation.

Comment (Xaver-DaRed) [reply scripts/enum/mob_mod.lua]: I would say those are 2 different factors/variables altoguether then.

However, why cant Korrigans and Tauris have their damage be "correct" (read: just have the base damage be the value expected, with the offset already applied) from the start?

The way I see it, it just cant be that we have base damage affected sometimes before and sometimes after the multiplier, that makes 0 sense. Stuff doesnt just change places at random.

Comment (UmeboshiXI) [reply scripts/enum/mob_mod.lua]: Removed the mobmods and overrides.

Korrigan and Tauri will be implemented by subtracting their offset so its 0 and then adding the value subtracted to WEAPON_BONUS.

Comment (Xaver-DaRed) [reply scripts/enum/mob_mod.lua]: /bow
tyvm
I wasnt trying to give you even more work, I'm just trying to make this maintainable and sane, while also accurate.

**Review by Xaver-DaRed — APPROVED** [review]

**Review by WinterSolstice8 — APPROVED** [review]

File: src/map/utils/mobutils.cpp [hunk 3]
```diff
@@ -55,30 +55,155 @@ namespace mobutils
...
         }
 
-        damage = lvl + bonus;
-
-        damage = (uint16)(damage * PMob->m_dmgMult / 100.0f);
+        // Clamp to 0 for edge cases that might cause the offset go negative.
+        if (offset < 0)
+        {
+            offset = 0;
+        }
```
Comment (Xaver-DaRed) [inline src/map/utils/mobutils.cpp:173]: Arent this 2 checks out of order?
imagen

1st, under X conditions, you are setting offset to be < 0
Then you are clamping it
